packages/odyssey-react-mui/src/BackgroundContext.tsx

```diff
@@ -0,0 +1,142 @@
+import * as Tokens from "@okta/odyssey-design-tokens";
+
+declare module "@mui/material/styles" {
+  interface Theme {
+    custom: {
+      isLowContrast: boolean;
+    };
+    odysseyContrastMode: ContrastMode;
+  }
+  interface ThemeOptions {
+    custom?: {
+      isLowContrast?: boolean;
+    };
+    odysseyContrastMode?: ContrastMode;
+  }
+}
+
+export type ContrastMode = "lowContrast" | "highContrast";
+
+export type BackgroundContextType = {
+  contrastMode: ContrastMode;
+  parentBackgroundColor: string;
+};
+
+const BackgroundContext = createContext<BackgroundContextType>({
+  contrastMode: "highContrast",
+  parentBackgroundColor: "",
+});
+
+export const useBackground = () => useContext(BackgroundContext);
+
+const hexToRgb = (hex: string): string => {
+  const bigint = parseInt(hex.slice(1), 16);
+  const r = (bigint >> 16) & 255;
+  const g = (bigint >> 8) & 255;
+  const b = bigint & 255;
+  return `rgb(${r}, ${g}, ${b})`;
+};
+
+export const useParentBackgroundColor = (ref: React.RefObject<HTMLElement>) => {
+  const [backgroundColor, setBackgroundColor] = useState("");
+
+  const hueNeutral50Rgb = useMemo(() => hexToRgb(Tokens.HueNeutral50), []);
+
+  useLayoutEffect(() => {
+    if (ref.current) {
+      let element: HTMLElement | null = ref.current;
+      while (element) {
+        const bgColor = window.getComputedStyle(element).backgroundColor;
+
+        if (bgColor !== "rgba(0, 0, 0, 0)" && bgColor !== "transparent") {
+          // Compare computed color to precomputed HueNeutral50 RGB value
+          if (bgColor === hueNeutral50Rgb) {
+            setBackgroundColor(Tokens.HueNeutral50);
+          } else {
+            setBackgroundColor(bgColor);
+          }
+          break;
+        }
+        element = element.parentElement;
+      }
+    }
+  }, [ref, hueNeutral50Rgb]);
+
+  return backgroundColor;
+};
+
+type BackgroundProviderProps = {
+  children: ReactNode;
+  contrastMode?: ContrastMode;
+};
+
+export const BackgroundProvider = ({
+  children,
+  contrastMode: explicitContrastMode,
+}: BackgroundProviderProps) => {
+  const ref = useRef<HTMLDivElement>(null);
+  const parentBackgroundColor = useParentBackgroundColor(ref);
+  const [contrastMode, setContrastMode] =
+    useState<ContrastMode>("highContrast");
+
+  useLayoutEffect(() => {
+    if (explicitContrastMode) {
+      setContrastMode(explicitContrastMode);
+    } else {
+      const isLowContrast = parentBackgroundColor === Tokens.HueNeutral50;
+      setContrastMode(isLowContrast ? "lowContrast" : "highContrast");
+    }
+  }, [parentBackgroundColor, explicitContrastMode]);
+
+  const contextValue = useMemo<BackgroundContextType>(
+    () => ({
+      contrastMode,
+      parentBackgroundColor,
+    }),
+    [contrastMode, parentBackgroundColor],
+  );
+
+  const existingTheme = useTheme();
+  const theme = useMemo(
+    () =>
+      createTheme({
+        ...existingTheme,
+        custom: {
```

Comment:
Do we want to name this key something more descriptive than just `custom`?

Reply:
We probably don't even need this top level object TBH. Any thoughts on just moving the contrast mode value to this level?

Reply:
I improved this and moved the contrast mode into the augmentation but I don't believe we can get rid of this entirely since it's used to pass the variables into the theme in OdysseyProvider so it can be used globally.